dot/digest/digest.go

```diff
@@ -267,23 +265,29 @@ func (h *Handler) persistBABEDigestsForNextEpoch(finalizedHeader *types.Header)
 	}
 
 	nextEpoch := currEpoch + 1
-	err = h.epochState.FinalizeBABENextEpochData(nextEpoch)
-	if err != nil && !errors.Is(err, state.ErrEpochNotInMemory) {
-		return fmt.Errorf("cannot finalize babe next epoch data for block number %d (%s): %w",
-			finalizedHeader.Number, finalizedHeader.Hash(), err)
+
+	appliedEpochData, appliedConfigData, err := h.epochState.AlreadyDefined(nextEpoch)
+	if err != nil {
+		return fmt.Errorf("cannot check if next epoch is already defined: %w", err)
 	}
 
-	err = h.epochState.FinalizeBABENextConfigData(nextEpoch)
-	if err == nil {
-		return nil
-	} else if errors.Is(err, state.ErrEpochNotInMemory) {
-		return fmt.Errorf("%w: %s", ErrDefineNextEpoch, err)
+	if !appliedEpochData {
+		err = h.epochState.FinalizeBABENextEpochData(nextEpoch)
+		if err != nil && !errors.Is(err, state.ErrEpochNotInMemory) {
```

Comment:
What do we do in the case where `errors.Is(err, state.ErrEpochNotInMemory)`? Should we just continue execution?

Reply:
I think this function should only be called if `finalizedHeader` is the first block in its epoch, as that's the only block that would contain `NextEpoch` or `NextConfig` digests. that would also prevent the spamming

Reply:
But, what about the case of the epoch 1 started at block `N` but the finalized block is the block `N + 2`, we will not persist the next epoch data because the finalized block is not the first in its epoch.

Reply:
right, yeah this should be fine then